### PR TITLE
Tuning of iterative Tracking at HLT using Cellular Automaton

### PR DESCRIPTION
**Pull request by JanFSchulte**
This PR is a cumulative update of the customization function for the iterative tracking to use CA seeing at HLT. 

Main features of this update: 
- Switches iterations 1 and 2 to use CA Quadruplets and Triplets, respectively
- Switches from CkfTrajectoryBuilder to GroupedCkfTrajectoryBuilder for all iterations, as promised at the last TSG workshop in the beginning of November, but so far not actually put into the release
- Tuning of the configuration in 9_0_0_pre4 to maximize performance

Detailed performance comparisons can be found here: https://cernbox.cern.ch/index.php/s/iF6QrRMTcqxPyB8
Compared are the performance of the new configuration to what is in pre4 right now, and also to the 2016 performance for reference. 

In summary, efficiency is significantly increased for no cost in fake rate. Timing improves from 220ms for 2016 to about 180ms for the new configuration, a speedup of about 20%. Note that this timing measurements have been done in 8_1_0_pre16 because the FastTimerService works in that release.

### CODE REVIEW

**Comment by Martin-Grunewald** [comment]
please test

File: HLTrigger/Configuration/python/customizeHLTTrackingForPhaseI2017.py [hunk 1]
```diff
@@ -106,130 +106,268 @@ def customizeHLTForPFTrackingPhaseI2017(process):
...
-        MeasurementTrackerName = cms.string( "hltIter1ESPMeasurementTracker" ),
-        estimator = cms.string( "hltESPChi2ChargeMeasurementEstimator16" ),
-        TTRHBuilder = cms.string( "hltESPTTRHBWithTrackAngle" ),
-        updator = cms.string( "hltESPKFUpdator" ),
-        alwaysUseInvalidHits = cms.bool( False ),
-        intermediateCleaning = cms.bool( True ),
-        lostHitPenalty = cms.double( 30.0 ),
-        useSameTrajFilter = cms.bool(False) # new ! other iteration should have it set to True
+    process.HLTIter1PSetTrajectoryFilterInOutIT = cms.PSet
+	ComponentType = cms.string('CkfBaseTrajectoryFilter'),
```
Comment (Martin-Grunewald) [inline HLTrigger/Configuration/python/customizeHLTTrackingForPhaseI2017.py]: There is a "(" missing on the line with cms.PSet.
It is good practice to actually test you code before making a PR.

Comment (JanFSchulte) [inline HLTrigger/Configuration/python/customizeHLTTrackingForPhaseI2017.py]: Sorry, made the usual mistake of "just" fixing the indentation to clean up the comparison...

**Comment by Martin-Grunewald** [comment]
please test

File: HLTrigger/Configuration/python/customizeHLTTrackingForPhaseI2017.py [hunk 2]
```diff
@@ -106,130 +106,268 @@ def customizeHLTForPFTrackingPhaseI2017(process):
...
+
+    process.HLTIterativeTrackingIteration1 = cms.Sequence( process.hltIter1ClustersRefRemoval + process.hltIter1MaskedMeasurementTrackerEvent + process.hltIter1PixelLayerQuadruplets + process.hltIter1PFlowPixelTrackingRegions + process.hltIter1PFlowPixelClusterCheck + process.hltIter1PFlowPixelHitDoublets + process.hltIter1PFlowPixelHitQuadruplets + process.hltIter1PFlowPixelSeeds + process.hltIter1PFlowCkfTrackCandidates + process.hltIter1PFlowCtfWithMaterialTracks + process.hltIter1PFlowTrackCutClassifierPrompt + process.hltIter1PFlowTrackCutClassifierDetached + process.hltIter1PFlowTrackCutClassifierMerged + process.hltIter1PFlowTrackSelectionHighPurity )
```
Comment (makortel) [inline HLTrigger/Configuration/python/customizeHLTTrackingForPhaseI2017.py]: I'd suggest to continue with `replace_with()`.

File: HLTrigger/Configuration/python/customizeHLTTrackingForPhaseI2017.py [hunk 3]
```diff
@@ -281,49 +427,89 @@ def _copy(old, new, skip=[]):
             if key not in skipSet:
                 setattr(new, key, getattr(old, key))
     from RecoTracker.TkSeedGenerator.seedCreatorFromRegionConsecutiveHitsTripletOnlyEDProducer_cfi import seedCreatorFromRegionConsecutiveHitsTripletOnlyEDProducer as _seedCreatorFromRegionConsecutiveHitsTripletOnlyEDProducer
-    replace_with(process.hltIter2PFlowPixelSeeds, _seedCreatorFromRegionConsecutiveHitsTripletOnlyEDProducer.clone(seedingHitSets="hltIter2PFlowPixelHitTriplets"))
+    process.hltIter2PFlowPixelSeeds = _seedCreatorFromRegionConsecutiveHitsTripletOnlyEDProducer.clone(seedingHitSets="hltIter2PFlowPixelHitTriplets")
```
Comment (makortel) [inline HLTrigger/Configuration/python/customizeHLTTrackingForPhaseI2017.py]: `replace_with()`

File: HLTrigger/Configuration/python/customizeHLTTrackingForPhaseI2017.py [hunk 4]
```diff
@@ -281,49 +427,89 @@ def _copy(old, new, skip=[]):
...
+	        useHitsSplitting = cms.bool( False ),
+	        doSeedingRegionRebuilding = cms.bool( False )
+    )	
+
+    process.hltIter2PFlowCkfTrackCandidates.TrajectoryBuilder = cms.string('HLTIter2GroupedCkfTrajectoryBuilderIT')
+    process.hltIter2PFlowCkfTrackCandidates.TrajectoryBuilderPSet = cms.PSet(refToPSet_ = cms.string('HLTIter2GroupedCkfTrajectoryBuilderIT'))
+
+
+
+    process.HLTIterativeTrackingIteration2 = cms.Sequence( process.hltIter2ClustersRefRemoval + process.hltIter2MaskedMeasurementTrackerEvent + process.hltIter2PixelLayerTriplets + process.hltIter2PFlowPixelTrackingRegions + process.hltIter2PFlowPixelClusterCheck + process.hltIter2PFlowPixelHitDoublets + process.hltIter2PFlowPixelHitTriplets + process.hltIter2PFlowPixelSeeds + process.hltIter2PFlowCkfTrackCandidates + process.hltIter2PFlowCtfWithMaterialTracks + process.hltIter2PFlowTrackCutClassifier + process.hltIter2PFlowTrackSelectionHighPurity )
```
Comment (makortel) [inline HLTrigger/Configuration/python/customizeHLTTrackingForPhaseI2017.py]: `replace_with`

File: HLTrigger/Configuration/python/customizeHLTTrackingForPhaseI2017.py [hunk 5]
```diff
@@ -281,49 +427,89 @@ def _copy(old, new, skip=[]):
...
+
+    # Need to operate on Paths as well...
+    for seqs in [process.sequences_(), process.paths_()]:
```
Comment (makortel) [inline HLTrigger/Configuration/python/customizeHLTTrackingForPhaseI2017.py]: Please keep the loop from #17412.

File: HLTrigger/Configuration/python/customizeHLTTrackingForPhaseI2017.py [hunk 6]
```diff
@@ -281,49 +427,89 @@ def _copy(old, new, skip=[]):
...
+                seq.insert(index, process.hltPixelTracksHitQuadruplets)
 
     # Remove entirely to avoid warning from the early deleter
     del process.hltPixelTracksHitTriplets
 
-    return process
+    for producer in producers_by_type(process,"PixelTripletHLTEDProducer"):
+        if "hltIter1PFlowPixelHitTriplets" in producer.label():
+
+    	   	for seqs in [process.sequences_(), process.paths_()]:
```
Comment (makortel) [inline HLTrigger/Configuration/python/customizeHLTTrackingForPhaseI2017.py]: Could the logic be included in the previous loop over all sequences, paths, and endpaths? (maybe collect beforehand the set of `PixelTripletHLTEDProducer`'s and do the necessary replacements for them.

File: HLTrigger/Configuration/python/customizeHLTTrackingForPhaseI2017.py [hunk 7]
```diff
@@ -106,130 +106,268 @@ def customizeHLTForPFTrackingPhaseI2017(process):
...
+            ComponentName = cms.string( "LowPtClusterShapeSeedComparitor" ),
+            clusterShapeCacheSrc = cms.InputTag( "hltSiPixelClustersCache" )
+        )
+
+
+    )
+
+
+ 
+    process.hltIter1PFlowPixelSeeds = cms.EDProducer( "SeedGeneratorFromProtoTracksEDProducer",
```
Comment (Martin-Grunewald) [inline HLTrigger/Configuration/python/customizeHLTTrackingForPhaseI2017.py]: Seconded! It will be a pain otherwise!

**Comment by JanFSchulte** [comment]
This latest commit incorporates the comments by Matti. Slides with plots are updated:
https://cernbox.cern.ch/index.php/s/iF6QrRMTcqxPyB8

File: HLTrigger/Configuration/python/customizeHLTTrackingForPhaseI2017.py [hunk 8]
```diff
@@ -106,130 +106,245 @@ def customizeHLTForPFTrackingPhaseI2017(process):
...
+     		propagatorOpposite = cms.string('PropagatorWithMaterialParabolicMfOpposite'),
+	        minNrOfHitsForRebuild = cms.int32(5),
+	        maxDPhiForLooperReconstruction = cms.double(2.0),
+	        maxPtForLooperReconstruction = cms.double(0.7),
+	        cleanTrajectoryAfterInOut = cms.bool( False ),
+	        useHitsSplitting = cms.bool( False ),
+	        doSeedingRegionRebuilding = cms.bool( False )
+    )
+
+    process.hltIter0PFlowCkfTrackCandidates.TrajectoryBuilder = cms.string('')
```
Comment (makortel) [inline HLTrigger/Configuration/python/customizeHLTTrackingForPhaseI2017.py]: I had in mind to remove this line (as it's a no-op), but ok.

File: HLTrigger/Configuration/python/customizeHLTTrackingForPhaseI2017.py [hunk 9]
```diff
@@ -106,130 +106,268 @@ def customizeHLTForPFTrackingPhaseI2017(process):
...
+	lostHitPenalty = cms.double( 30.0 ),
+	useSameTrajFilter = cms.bool(False) 
     )
+    process.HLTIter1GroupedCkfTrajectoryBuilderIT = cms.PSet(
+	ComponentType = cms.string('GroupedCkfTrajectoryBuilder'),
+	bestHitOnly = cms.bool(True),
+	propagatorAlong = cms.string('PropagatorWithMaterialParabolicMf'),
+	trajectoryFilter = cms.PSet(refToPSet_ = cms.string('HLTIter1PSetTrajectoryFilterIT')),
+	inOutTrajectoryFilter = cms.PSet(refToPSet_ = cms.string('HLTIter1PSetTrajectoryFilterInOutIT')),
+	useSameTrajFilter = cms.bool(False),
```
Comment (makortel) [inline HLTrigger/Configuration/python/customizeHLTTrackingForPhaseI2017.py]: Are the two trajectory filters now exactly the same (I didn't check)? If yes, it would be more clear to set `useSameTrajFilter=True` and use same value for `inOutTrajectoryFilter` as for `trajectoryFilter`. If no, why?

File: HLTrigger/Configuration/python/customizeHLTTrackingForPhaseI2017.py (hunk 3 above)
Comment (makortel) [inline HLTrigger/Configuration/python/customizeHLTTrackingForPhaseI2017.py]: Please use `replace_with()`.

File: HLTrigger/Configuration/python/customizeHLTTrackingForPhaseI2017.py [hunk 10]
```diff
@@ -281,11 +404,44 @@ def _copy(old, new, skip=[]):
...
+	        propagatorOpposite = cms.string('PropagatorWithMaterialParabolicMfOpposite'),
+	        minNrOfHitsForRebuild = cms.int32(5),
+	        maxDPhiForLooperReconstruction = cms.double(2.0),
+	        maxPtForLooperReconstruction = cms.double(0.7),
+	        cleanTrajectoryAfterInOut = cms.bool( False ),
+	        useHitsSplitting = cms.bool( False ),
+	        doSeedingRegionRebuilding = cms.bool( False )
+    )	
+
+    process.hltIter2PFlowCkfTrackCandidates.TrajectoryBuilder = cms.string('')
```
Comment (makortel) [inline HLTrigger/Configuration/python/customizeHLTTrackingForPhaseI2017.py]: I'd remove this line as it's a no-op.

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by Martin-Grunewald** [comment]
+1

**Comment by davidlange6** [comment]
+1
